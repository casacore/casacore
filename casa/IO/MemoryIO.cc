//# MemoryIO.cc: Class for IO in memory
//# Copyright (C) 1996,1999,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$


#include <casa/IO/MemoryIO.h>
#include <casa/Utilities/Assert.h>
#include <casa/Exceptions/Error.h>
#include <cstring>                  //# for memcpy with gcc-4.3

namespace casa { //# NAMESPACE CASA - BEGIN

MemoryIO::MemoryIO (uInt64 initialSize, uInt64 expandSize)
: itsBuffer     (0),
  itsAlloc      (initialSize),
  itsExpandSize (expandSize),
  itsUsed       (0),
  itsPosition   (0),
  itsReadable   (True),
  itsWritable   (True),
  itsCanDelete  (True)
{
  if (itsAlloc > 0) {
    itsBuffer = new uChar[itsAlloc];
    AlwaysAssert (itsBuffer != 0, AipsError);
  }
}

MemoryIO::MemoryIO (const void* buffer, uInt64 size)
: itsBuffer     ((uChar*)buffer),
  itsAlloc      (size),
  itsExpandSize (0),
  itsUsed       (size),
  itsPosition   (0),
  itsReadable   (True),
  itsWritable   (False),
  itsCanDelete  (False)
{}

MemoryIO::MemoryIO (void* buffer, uInt64 size, ByteIO::OpenOption option,
		    uInt64 expandSize, Bool canDelete)
: itsBuffer     ((uChar*)buffer),
  itsAlloc      (size),
  itsExpandSize (expandSize),
  itsUsed       (size),
  itsPosition   (0),
  itsReadable   (True),
  itsWritable   (True),
  itsCanDelete  (canDelete)
{
  // Make sure there is a buffer.
  if (itsAlloc > 0) {
    AlwaysAssert (itsBuffer != 0, AipsError);
  }
  // Adapt position, etc. from the option.
  switch (option) {
  case ByteIO::Old:
    itsWritable = False;
    break;
  case ByteIO::Append:
    itsPosition = itsUsed;
    break;
  default:
    itsUsed = 0;
    break;
  }
}


MemoryIO::~MemoryIO()
{
  if (itsCanDelete) {
    delete [] itsBuffer;
  }
}

void MemoryIO::write (uInt size, const void* buf)
{
  // Throw an exception if not writable.
  if (!itsWritable) {
    throw (AipsError ("MemoryIO::write - MemoryIO object is not writable"));
  }
  // Expand the buffer when needed (and possible).
  Int64 minSize = itsPosition + size;
  if (minSize > itsAlloc) {
    if (! expand (minSize)) {
      throw (AipsError ("MemoryIO::write - buffer cannot be expanded"));
    }
  }
  // Copy the data and set new position and used.
  memcpy (itsBuffer + itsPosition, buf, size);
  itsPosition += size;
  if (itsPosition > itsUsed) {
    itsUsed = itsPosition;
  }
}

Int MemoryIO::read (uInt size, void* buf, Bool throwException)
{
  // Throw an exception if not readable.
  if (!itsReadable) {
    throw (AipsError ("MemoryIO::read - buffer is not readable"));
  }
  const Int64 bytesLeft = itsUsed - itsPosition;
  Int bytesRead = 0;
  if (Int(size) <= bytesLeft) {
    memcpy (buf, itsBuffer + itsPosition, size);
    itsPosition += size;
    bytesRead = size;
  } else {
    if (bytesLeft >= 0) { 
      bytesRead = bytesLeft;
      memcpy (buf, itsBuffer + itsPosition, bytesRead);
      itsPosition += bytesLeft;
      if (throwException) {
	throw (AipsError ("MemoryIO::read - incorrect number of bytes read"));
      }
    } else {
      throw (AipsError ("MemoryIO::read - buffer position is invalid"));
    }
  }
  return bytesRead;
}

Int64 MemoryIO::doSeek (Int64 offset, ByteIO::SeekOption dir)
{
  // Determine the new position.
  // Exit with error status if negative.
  Int64 newPos;
  switch (dir) {
  case ByteIO::Begin:
    newPos = offset;
    break;
  case ByteIO::End:
    newPos = itsUsed + offset;
    break;
  default:
    newPos = itsPosition + offset;
    break;
  }
  if (newPos < 0) {
    throw (AipsError("MemoryIO::seek - cannot seek before start of object"));
  }
  // It is possible to seek past the end of the buffer.
  // This means that the buffer usage increases.
  // Expand the buffer if needed.
  // Initialize the new buffer positions with zeroes.
  if (newPos > itsUsed) {
    // Throw an exception if not writable.
    if (!itsWritable) {
      throw (AipsError ("MemoryIO::seek - cannot seek past the end "
			"of a readonly object"));
    }
    if (newPos > itsAlloc) {
      if (! expand (newPos)) {
	throw (AipsError("MemoryIO::seek - buffer cannot be expanded"));
      }
    }
    for (; itsUsed<newPos; itsUsed++) {
      itsBuffer[itsUsed] = 0;
    }
  }
  itsPosition = newPos;
  return newPos;
}


Bool MemoryIO::expand (uInt64 minSize)
{
  Int64 minsz = minSize;
  // Check if expansion is really needed.
  if (minsz <= itsAlloc) {
    return True;
  }
  // Return with error status when expansion is not possible.
  if (itsExpandSize == 0) {
    return False;
  }
  // Expand with at least the expansion size.
  if (minsz < itsAlloc + itsExpandSize) {
    minsz = itsAlloc + itsExpandSize;
  }
  // Allocate new buffer, copy contents and delete old buffer (if possible).
  uChar* newBuffer = new uChar[minsz];
  AlwaysAssert (newBuffer != 0, AipsError);
  // Copy the old contents (if any).
  if (itsBuffer != 0) {
    memcpy (newBuffer, itsBuffer, itsUsed);
    if (itsCanDelete) {
      delete [] itsBuffer;
    }
  }
  itsBuffer    = newBuffer;
  itsAlloc     = minsz;
  itsCanDelete = True;
  return True;
}


Int64 MemoryIO::length()
{
  return itsUsed;
}

Bool MemoryIO::isReadable() const
{
  return itsReadable;
}
Bool MemoryIO::isWritable() const
{
  return itsWritable;
}
Bool MemoryIO::isSeekable() const
{
  return True;
}

void MemoryIO::setUsed (uInt64 bytesUsed)
{
  if (!itsWritable) {
    throw (AipsError ("MemoryIO::setUsed - object is not writable"));
  }
  if (Int64(bytesUsed) > itsAlloc) {
    throw(AipsError ("MemoryIO::setUsed - cannot use more than is allocated"));
  }
  itsUsed = bytesUsed;
}

uChar* MemoryIO::setBuffer (uInt64 length)
{
  if (!itsWritable) {
    throw (AipsError ("MemoryIO::setBuffer - object is not writable"));
  }
  if (!expand (length)) {
    throw(AipsError ("MemoryIO::setBuffer - buffer cannot be expanded"));
  }
  itsUsed = length;
  return itsBuffer;
}

} //# NAMESPACE CASA - END

