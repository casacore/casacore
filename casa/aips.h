//# aips.h: Global initialization for standard types, etc.
//# Copyright (C) 1993-1999,2000,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_AIPS_H
#define CASA_AIPS_H

//# Define compiler specific flags
#include <casacore/casa/aipsenv.h>

//# Define the standard types used by Casacore.
#include <casacore/casa/aipstype.h>

//# Define the extra non-standard types used by Casacore
#include <casacore/casa/aipsxtype.h>

//# Define the special Casacore macros
#include <casacore/casa/aipsdef.h>

namespace casa { //# NAMESPACE CASA - BEGIN

} //# NAMESPACE CASA - END

#endif
