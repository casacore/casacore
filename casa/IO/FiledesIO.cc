//# FiledesIO.cc: Class for IO on a file descriptor
//# Copyright (C) 1997,1999,2001
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casa/aips.h>
#include <casa/IO/FiledesIO.h>
#include <casa/Utilities/Assert.h>
#include <casa/Exceptions/Error.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>                // needed for errno
#include <casa/string.h>               // needed for strerror

#ifdef PABLO_IO
#include "IOTrace.h"
#else
#define traceFCLOSE fclose
#define traceFSEEK fseek
#define traceFREAD fread
#define traceFWRITE fwrite
#define traceREAD read
#define traceWRITE write
#define trace2OPEN open
#define traceLSEEK lseek
#define trace3OPEN open
#define traceCLOSE close
#endif // PABLO_IO

namespace casa { //# NAMESPACE CASA - BEGIN

FiledesIO::FiledesIO()
: itsSeekable (False),
  itsReadable (False),
  itsWritable (False),
  itsFile     (-1)
{}

FiledesIO::FiledesIO (int fd)
: itsFile  (-1)
{
    attach (fd);
}

FiledesIO::~FiledesIO()
{
    detach();
}


void FiledesIO::attach (int fd)
{
    AlwaysAssert (itsFile == -1, AipsError);
    itsFile = fd;
    fillRWFlags (fd);
    fillSeekable();
}

void FiledesIO::detach()
{
    itsFile = -1;
}

void FiledesIO::fillRWFlags (int fd)
{
    itsReadable = False;
    itsWritable = False;
    int flags = fcntl (fd, F_GETFL);
    if ((flags & O_RDWR)  ==  O_RDWR) {
	itsReadable = True;
	itsWritable = True;
    } else if ((flags & O_WRONLY)  ==  O_WRONLY) {
	itsWritable = True;
    } else {
	itsReadable = True;
    }
}

void FiledesIO::fillSeekable()
{
    itsSeekable = (seek (0, ByteIO::Current)  >= 0);
}


String FiledesIO::fileName() const
{
    return "";
}


void FiledesIO::write (uInt size, const void* buf)
{
    // Throw an exception if not writable.
    if (!itsWritable) {
	throw (AipsError ("FiledesIO object is not writable"));
    }
    if (::traceWRITE(itsFile, (Char *)buf, size) != Int(size)) {
	throw (AipsError (String("FiledesIO: write error: ")
			  + strerror(errno)));
    }
}

Int FiledesIO::read (uInt size, void* buf, Bool throwException)
{
  // Throw an exception if not readable.
  if (!itsReadable) {
    throw (AipsError ("FiledesIO::read - descriptor is not readable"));
  }
  Int bytesRead = ::traceREAD (itsFile, (Char *)buf, size);
  if (bytesRead > Int(size)) { // Should never be executed
    throw (AipsError ("FiledesIO::read - read returned a bad value"));
  }
  if (bytesRead != Int(size) && throwException == True) {
    if (bytesRead < 0) {
      throw (AipsError (String("FiledesIO::read - "
			       " error returned by system call: ") + 
			strerror(errno)));
    } else if (bytesRead < Int(size)) {
      throw (AipsError ("FiledesIO::read - incorrect number of bytes read"));
    }
  }
  return bytesRead;
}

Int64 FiledesIO::doSeek (Int64 offset, ByteIO::SeekOption dir)
{
    switch (dir) {
    case ByteIO::Begin:
	return ::traceLSEEK (itsFile, offset, SEEK_SET);
    case ByteIO::End:
        return ::traceLSEEK (itsFile, offset, SEEK_END);
    default:
	break;
    }
    return ::traceLSEEK (itsFile, offset, SEEK_CUR);
}

Int64 FiledesIO::length()
{
    // Get current position to be able to reposition.
    Int64 pos = seek (0, ByteIO::Current);
    // Seek to the end of the stream.
    // If it fails, we cannot seek and the current position is the length.
    Int64 len = seek (0, ByteIO::End);
    if (len < 0) {
	return pos;
    }
    // Reposition and return the length.
    seek (pos, ByteIO::Begin);
    return len;
}

   
Bool FiledesIO::isReadable() const
{
    return itsReadable;
}

Bool FiledesIO::isWritable() const
{
    return itsWritable;
}

Bool FiledesIO::isSeekable() const
{
    return itsSeekable;
}


int FiledesIO::create (const Char* name, int mode)
{
    int fd = ::trace3OPEN ((Char *)name, O_RDWR | O_CREAT | O_TRUNC, mode);
    if (fd == -1) {
	throw (AipsError ("FiledesIO: file " + String(name) +
			  " could not be created: " + strerror(errno)));
    }
    return fd;
}
int FiledesIO::open (const Char* name, Bool writable, Bool throwExcp)
{
    int fd;
    if (writable) {
	fd = ::trace2OPEN ((Char *)name, O_RDWR);
    }else{
	fd = ::trace2OPEN ((Char *)name, O_RDONLY);
    }
    if (throwExcp  &&  fd == -1) {
	throw (AipsError ("FiledesIO: file " + String(name) +
			  " could not be opened: " + strerror(errno)));
    }
    return fd;
}
void FiledesIO::close (int fd)
{
    if (::traceCLOSE (fd)  == -1) {
	throw (AipsError (String("FiledesIO: file could not be closed: ")
			  + strerror(errno)));
    }
}

} //# NAMESPACE CASA - END

