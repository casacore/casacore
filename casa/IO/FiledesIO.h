//# FiledesIO.h: Class for IO on a file descriptor
//# Copyright (C) 1996,1997,1999,2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#ifndef CASA_FILEDESIO_H
#define CASA_FILEDESIO_H

//# Includes
#include <casa/aips.h>
#include <casa/IO/ByteIO.h>
#include <casa/BasicSL/String.h>

namespace casa { //# NAMESPACE CASA - BEGIN

// <summary>Class for IO on a file descriptor.</summary>

// <use visibility=export>

// <reviewed reviewer="Friso Olnon" date="1996/11/06" tests="tByteIO" demos="">
// </reviewed>

// <prerequisite> 
//    <li> <linkto class=ByteIO>ByteIO</linkto> class
//    <li> file descriptors
// </prerequisite>

// <synopsis> 
// This class is a specialization of class
// <linkto class=ByteIO>ByteIO</linkto>. It uses a file descriptor
// to read/write data.
// <p>
// The file associated with the file descriptor has to be opened
// before hand.
// The constructor will determine automatically if the file is
// readable, writable and seekable.
// Note that on destruction the file descriptor is NOT closed.
// </synopsis>

// <example>
// This example shows how FiledesIO can be used with an fd.
// It uses the fd for a regular file, which could be done in an easier
// way using class <linkto class=RegularFileIO>RegularFileIO</linkto>.
// However, when using pipes or sockets, this would be the only way.
// <srcblock>
//    // Get a file descriptor for the file.
//    int fd = open ("file.name");
//    // Use that as the source of AipsIO (which will also use CanonicalIO).
//    FiledesIO fio (fd);
//    AipsIO stream (&fio);
//    // Read the data.
//    Int vali;
//    Bool valb;
//    stream >> vali >> valb;
// </srcblock>
// </example>

// <motivation> 
// Make it possible to use the Casacore IO functionality on any file.
// In this way any device can be hooked to the IO framework.
// </motivation>


class FiledesIO: public ByteIO
{
public: 
    // Default constructor.
    // A stream can be attached using the attach function.
    FiledesIO();

    // Construct from the given file descriptor.
    // The file name is only used in possible error messages.
    explicit FiledesIO (int fd, const String& fileName=String());

    // Attach to the given file descriptor.
    // An exception is thrown if it is not in a detached state.
    // The file name is only used in error messages.
    void attach (int fd, const String& fileName);

    // The destructor does not close the file.
    ~FiledesIO();
    
    // Write the number of bytes.
    virtual void write (uInt size, const void* buf);

    // Read <src>size</src> bytes from the descriptor. Returns the number of
    // bytes actually read or a negative number if an error occured. Will throw
    // an Exception (AipsError) if the requested number of bytes could not be
    // read, or an error occured, unless throwException is set to False. Will
    // always throw an exception if the descriptor is not readable or the
    // system call returned an undocumented value.
    virtual Int read (uInt size, void* buf, Bool throwException=True);    

    // Get the length of the byte stream.
    virtual Int64 length();
       
    // Is the IO stream readable?
    virtual Bool isReadable() const;

    // Is the IO stream writable?
    virtual Bool isWritable() const;

    // Is the IO stream seekable?
    virtual Bool isSeekable() const;

    // Get the file name of the file attached.
    const String& fileName() const
      { return itsFileName; }

    // Some static convenience functions for file create/open/close.
    // <group>
    static int create (const Char* name, int mode = 0666);
    static int open   (const Char* name, Bool writable = False,
		       Bool throwExcp = True);
    static void close (int fd);
    // </group>


protected:
    // Get the file descriptor.
    int fd() const
      { return itsFile; }

    // Detach from the file descriptor. It is not closed.
    void detach();

    // Determine if the file descriptor is readable and/or writable.
    void fillRWFlags (int fd);

    // Determine if the file is seekable.
    void fillSeekable();

    // Reset the position pointer to the given value. It returns the
    // new position.
    virtual Int64 doSeek (Int64 offset, ByteIO::SeekOption);

private:
    Bool   itsSeekable;
    Bool   itsReadable;
    Bool   itsWritable;
    int    itsFile;
    String itsFileName;

    // Copy constructor, should not be used.
    FiledesIO (const FiledesIO& that);

    // Assignment, should not be used.
    FiledesIO& operator= (const FiledesIO& that);
};




} //# NAMESPACE CASA - END

#endif
