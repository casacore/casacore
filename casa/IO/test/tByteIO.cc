//# tByteIO.cc: Test program for class ByteIO and derived classes
//# Copyright (C) 1996,1997,2000,2001,2002
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id$

#include <casa/aips.h>
#include <casa/IO/FiledesIO.h>
#include <casa/IO/RegularFileIO.h>
#include <casa/IO/MemoryIO.h>
#include <casa/OS/RegularFile.h>
#include <casa/Utilities/Assert.h>
#include <casa/Exceptions/Error.h>
#include <casa/iostream.h>
#include <unistd.h>
#include <fcntl.h>


#include <casa/namespace.h>
void checkLength (ByteIO& fio, uInt& curLength, uInt addLength)
{
    curLength += addLength;
    AlwaysAssertExit (fio.length() == curLength);
}


static Bool valb=True;
static Short vals=-3;
static uShort valus=2;
static Int vali=1000;
static uInt valui=32768;
static Int64 vall=-14793;
static uInt64 valul=17;
static float valf=1.2;
static double vald=-3.14;

void checkValues (ByteIO& fio, uShort incr)
{
    fio.seek (0);
    uInt curLength = fio.length();
    Bool resb;
    AlwaysAssertExit (fio.read (sizeof(Bool), &resb) == sizeof(Bool));
    AlwaysAssertExit (resb == valb);
    Short ress;
    AlwaysAssertExit (fio.read (sizeof(Short), &ress) == sizeof(Short));
    AlwaysAssertExit (ress == vals - incr);
    uShort resus;
    AlwaysAssertExit (fio.read (sizeof(uShort), &resus) == sizeof(uShort));
    AlwaysAssertExit (resus == valus + incr);
    Int resi;
    AlwaysAssertExit (fio.read (sizeof(Int), &resi) == sizeof(Int));
    AlwaysAssertExit (resi == vali);
    uInt resui;
    AlwaysAssertExit (fio.read (sizeof(uInt), &resui) == sizeof(uInt));
    AlwaysAssertExit (resui == valui);
    Int64 resl;
    AlwaysAssertExit (fio.read (sizeof(Int64), &resl) == sizeof(Int64));
    AlwaysAssertExit (resl == vall);
    uInt64 resul;
    AlwaysAssertExit (fio.read (sizeof(uInt64), &resul) == sizeof(uInt64));
    AlwaysAssertExit (resul == valul);
    float resf;
    AlwaysAssertExit (fio.read (sizeof(float), &resf) == sizeof(float));
    AlwaysAssertExit (resf == valf);
    double resd;
    AlwaysAssertExit (fio.read (sizeof(double), &resd) == sizeof(double));
    AlwaysAssertExit (resd == vald);
    AlwaysAssertExit (fio.length() == curLength);
}

void doIt (ByteIO& fio)
{
    uInt length = 0;
    AlwaysAssertExit (fio.length() == 0);
    fio.write (sizeof(Bool), &valb);
    checkLength (fio, length, sizeof(Bool));
    fio.write (sizeof(Short), &vals);
    checkLength (fio, length, sizeof(Short));
    fio.write (sizeof(uShort), &valus);
    checkLength (fio, length, sizeof(uShort));
    fio.write (sizeof(Int), &vali);
    checkLength (fio, length, sizeof(Int));
    fio.write (sizeof(uInt), &valui);
    checkLength (fio, length, sizeof(uInt));
    fio.write (sizeof(Int64), &vall);
    checkLength (fio, length, sizeof(Int64));
    fio.write (sizeof(uInt64), &valul);
    checkLength (fio, length, sizeof(uInt64));
    fio.write (sizeof(float), &valf);
    checkLength (fio, length, sizeof(float));
    fio.write (sizeof(double), &vald);
    checkLength (fio, length, sizeof(double));
    
    checkValues (fio, 0);
    
    fio.seek (Int(sizeof(Bool)));
    AlwaysAssertExit (fio.length() == length);
    uShort incr = 100;
    Short vals1 = vals - incr;
    Short ress;
    fio.write (sizeof(Short), &vals1);
    fio.seek (Int(sizeof(Bool)));
    AlwaysAssertExit (fio.read (sizeof(Short), &ress) == sizeof(Short));
    AlwaysAssertExit (ress == vals1);
    uShort valus1 = valus + incr;
    uShort resus;
    fio.write (sizeof(uShort), &valus1);
    fio.seek (Int(-sizeof(uShort)), ByteIO::Current);
    AlwaysAssertExit (fio.read (sizeof(uShort), &resus) == sizeof(uShort));
    AlwaysAssertExit (resus == valus1);
    Int resi;
    AlwaysAssertExit (fio.read (sizeof(Int), &resi) == sizeof(Int));
    AlwaysAssertExit (resi == vali);
    AlwaysAssertExit (fio.length() == length);
    
    checkValues (fio, incr);
}

void checkReopen()
{
    RegularFile rfile("tByteIO_tmp.data");
    {
	RegularFileIO fio(rfile);
	checkValues (fio, 100);
	fio.reopenRW();
	fio.seek (Int64(sizeof(Bool)));
	Short vals;
	fio.read (sizeof(Short), &vals);
	vals -= 50;
	fio.seek (Int64(sizeof(Bool)));
	fio.write (sizeof(Short), &vals);
	uShort valus;
	fio.read (sizeof(uShort), &valus);
	valus += 50;
	fio.seek (Int(-sizeof(uShort)), ByteIO::Current);
	fio.write (sizeof(uShort), &valus);
	checkValues (fio, 150);
    }

    rfile.setPermissions (0444);
    RegularFileIO fio2(rfile);
    checkValues (fio2, 150);
    Bool flag = False;
    try {
	fio2.reopenRW();
    } catch (AipsError x) {
	flag = True;
    } 
    AlwaysAssertExit (flag);
    checkValues (fio2, 150);
    rfile.setPermissions (0644);
}

void testMemoryIO()
{
    {
	uChar buf[10];
	MemoryIO membuf (buf, sizeof(buf), ByteIO::New, 6);
	doIt (membuf);
	AlwaysAssertExit (membuf.getBuffer() != (const uChar*)&buf);
	Int64 length = membuf.length();
	Int incr = 20;
	membuf.seek (incr, ByteIO::End);
	AlwaysAssertExit (membuf.length() == length+incr);
	checkValues (membuf, 100);
	char val;
	Int64 lincr = incr;
	membuf.seek (-lincr, ByteIO::End);
	for (Int i=0; i<incr; i++) {
	    membuf.read (1, &val);
	    AlwaysAssertExit (val == 0);
	}
	try {
	    membuf.read (1, &val);
	} catch (AipsError x) {
	    cout << x.getMesg() << endl;         // read beyond object
	} 
	try {
	    membuf.seek (Int(-(length + incr + 1)), ByteIO::Current);
	} catch (AipsError x) {
	    cout << x.getMesg() << endl;         // negative seek
	} 
	membuf.seek (Int(-(length + incr)), ByteIO::Current);
    }
    {
	char buf[10];
	MemoryIO membuf (buf, sizeof(buf), ByteIO::New, 0);
	try {
	    doIt (membuf);
	} catch (AipsError x) {                   // not expandable
	    cout << x.getMesg() << endl;
	} 
	try {
	    membuf.seek (10, ByteIO::End);
	} catch (AipsError x) {                   // not expandable
	    cout << x.getMesg() << endl;
	} 
	AlwaysAssertExit (membuf.getBuffer() == (const uChar*)buf);
    }
    {
	char* buf = new char[10];
	MemoryIO membuf (buf, 10, ByteIO::Scratch, 0, True);
	try {
	    doIt (membuf);
	} catch (AipsError x) {                   // not expandable
	    cout << x.getMesg() << endl;
	} 
	try {
	    membuf.seek (10, ByteIO::End);
	} catch (AipsError x) {                   // not expandable
	    cout << x.getMesg() << endl;
	} 
	AlwaysAssertExit (membuf.getBuffer() == (const uChar*)buf);
    }
}


int main()
{
    try {
	testMemoryIO();

	MemoryIO file2;
	doIt (file2);
	
	MemoryIO file3 (file2.getBuffer(), file2.length());
	try {
	    file3.write (0, 0);
	} catch (AipsError x) {
	    cout << x.getMesg() << endl;          // readonly
	} 
	checkValues (file3, 100);
	
	{
	    RegularFileIO file1(RegularFile("tByteIO_tmp.data"), ByteIO::New);
	    doIt (file1);
	}
	checkReopen();
	// Do regular io for various buffer sizes.
	for (uInt bs=1; bs<100; bs++) {
	    RegularFileIO file1(RegularFile("tByteIO_tmp.data"), ByteIO::New,
				bs);
	    doIt (file1);
	}

	int fd = open ("tByteIO_tmp.data2", O_CREAT|O_TRUNC|O_RDWR, 0644);
	int flags = fcntl (fd, F_GETFL);
	if (flags & O_RDWR) {
	    cout << "read/write" << endl;
	} else if (flags & O_WRONLY) {
	    cout << "writeonly" << endl;
	} else {
	    cout << "readonly" << endl;
	}
	FiledesIO file4 (fd, "");
	doIt (file4);
	close (fd);

	int fd1 = open ("tByteIO_tmp.data2", O_RDONLY, 0644);
	int flags1 = fcntl (fd1, F_GETFL);
	if (flags1 & O_RDWR) {
	    cout << "read/write" << endl;
	} else if (flags1 & O_WRONLY) {
	    cout << "writeonly" << endl;
	} else {
	    cout << "readonly" << endl;
	}
	FiledesIO file5 (fd1, "");
	checkValues (file5, 100);
	close (fd1);

	int fd2 = creat ("tByteIO_tmp.data2", 0644);
	int flags2 = fcntl (fd2, F_GETFL);
	if (flags2 & O_RDWR) {
	    cout << "read/write" << endl;
	} else if (flags2 & O_WRONLY) {
	    cout << "writeonly" << endl;
	} else {
	    cout << "readonly" << endl;
	}
	FiledesIO file6 (fd2, "");
	close (fd2);
    } catch (AipsError x) {
	cout << "Caught an exception: " << x.getMesg() << endl;
	return 1;
    } 
    return 0;                           // exit with success status
}

