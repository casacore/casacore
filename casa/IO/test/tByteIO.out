MemoryIO::read - incorrect number of bytes read
MemoryIO::seek - cannot seek before start of object
MemoryIO::write - buffer cannot be expanded
MemoryIO::seek - buffer cannot be expanded
MemoryIO::write - buffer cannot be expanded
MemoryIO::seek - buffer cannot be expanded
MemoryIO::write - MemoryIO object is not writable
read/write
readonly
writeonly
